internal/services/mongodb/instance.go

```diff
@@ -0,0 +1,405 @@
+		}
+	}
+	d.SetId(zonal.NewIDString(zone, res.ID))
+	_, err = waitForInstance(ctx, mongodbAPI, res.Region, res.ID, d.Timeout(schema.TimeoutCreate))
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	return ResourceInstanceRead(ctx, d, m)
+}
+
+func ResourceInstanceRead(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	mongodbAPI, region, ID, err := NewAPIWithRegionAndID(m, d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	getReq := &mongodb.GetInstanceRequest{
+		Region:     region,
+		InstanceID: ID,
+	}
+
+	instance, err := mongodbAPI.GetInstance(getReq, scw.WithContext(ctx))
+	if err != nil {
+		if httperrors.Is404(err) {
+			d.SetId("")
+			return nil
+		}
+		return diag.FromErr(err)
+	}
+
+	_ = d.Set("name", instance.Name)
+	_ = d.Set("version", instance.Version)
+	_ = d.Set("node_number", int(instance.NodeNumber))
+	_ = d.Set("node_type", instance.NodeType)
+	_ = d.Set("project_id", instance.ProjectID)
+	_ = d.Set("tags", instance.Tags)
+	_ = d.Set("created_at", instance.CreatedAt.Format(time.RFC3339))
+	_ = d.Set("region", instance.Region.String())
+
+	if instance.Volume != nil {
+		_ = d.Set("volume_type", instance.Volume.Type)
+		_ = d.Set("volume_size_in_gb", int(instance.Volume.Size/scw.GB))
+	}
+
+	publicNetworkEndpoint, publicNetworkExists := flattenPublicNetwork(instance.Endpoints)
+	if publicNetworkExists {
+		_ = d.Set("public_network", publicNetworkEndpoint)
+	}
+
+	if len(instance.Settings) > 0 {
+		settingsMap := make(map[string]string)
+		for _, setting := range instance.Settings {
+			settingsMap[setting.Name] = setting.Value
+		}
+		_ = d.Set("settings", settingsMap)
+	}
+
+	return nil
+}
+
+func ResourceInstanceUpdate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	mongodbAPI, region, ID, err := NewAPIWithRegionAndID(m, d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	////////////////////
+	// Upgrade instance
+	////////////////////
+
+	if d.HasChange("volume_size_in_gb") {
+		oldSizeInterface, newSizeInterface := d.GetChange("volume_size_in_gb")
+		oldSize := uint64(oldSizeInterface.(int))
+		newSize := uint64(newSizeInterface.(int))
+		if newSize < oldSize {
+			return diag.FromErr(errors.New("volume_size_in_gb cannot be decreased"))
+		}
+
+		if newSize%5 != 0 {
+			return diag.FromErr(errors.New("volume_size_in_gb must be a multiple of 5"))
+		}
+		size := scw.Size(newSize * uint64(scw.GB))
+
+		upgradeInstanceRequests := mongodb.UpgradeInstanceRequest{
+			InstanceID: ID,
+			Region:     region,
+			VolumeSize: &size,
+		}
+
+		_, err = mongodbAPI.UpgradeInstance(&upgradeInstanceRequests, scw.WithContext(ctx))
+		if err != nil {
+			return diag.FromErr(err)
+		}
+		_, err = waitForInstance(ctx, mongodbAPI, region, ID, d.Timeout(schema.TimeoutUpdate))
+		if err != nil {
+			return diag.FromErr(err)
+		}
+	}
+
+	req := &mongodb.UpdateInstanceRequest{
+		Region:     region,
+		InstanceID: ID,
+	}
+
+	if d.HasChange("name") {
```

Comment:
Introduce a variable that is changed like `should_update` or `has_changed` and if this variable is set then perform the update. That way the conditional line 342 won't get bigger